// react-draft-js-inline-toolbar-plugin/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Separator = exports.CodeBlockButton = exports.BlockquoteButton = exports.OrderedListButton = exports.UnorderedListButton = exports.HeaderThreeButton = exports.HeaderTwoButton = exports.HeaderOneButton = exports.UnderlineButton = exports.CodeButton = exports.BoldButton = exports.ItalicButton = undefined;

var _decorateComponentWithProps = require('decorate-component-with-props');

var _decorateComponentWithProps2 = _interopRequireDefault(_decorateComponentWithProps);

var _createStore = require('./utils/createStore');

var _createStore2 = _interopRequireDefault(_createStore);

var _Toolbar = require('./components/Toolbar');

var _Toolbar2 = _interopRequireDefault(_Toolbar);

var _ItalicButton = require('./components/ItalicButton');

var _ItalicButton2 = _interopRequireDefault(_ItalicButton);

var _BoldButton = require('./components/BoldButton');

var _BoldButton2 = _interopRequireDefault(_BoldButton);

var _CodeButton = require('./components/CodeButton');

var _CodeButton2 = _interopRequireDefault(_CodeButton);

var _UnderlineButton = require('./components/UnderlineButton');

var _UnderlineButton2 = _interopRequireDefault(_UnderlineButton);

var _HeaderOneButton = require('./components/HeaderOneButton');

var _HeaderOneButton2 = _interopRequireDefault(_HeaderOneButton);

var _HeaderTwoButton = require('./components/HeaderTwoButton');

var _HeaderTwoButton2 = _interopRequireDefault(_HeaderTwoButton);

var _HeaderThreeButton = require('./components/HeaderThreeButton');

var _HeaderThreeButton2 = _interopRequireDefault(_HeaderThreeButton);

var _UnorderedListButton = require('./components/UnorderedListButton');

var _UnorderedListButton2 = _interopRequireDefault(_UnorderedListButton);

var _OrderedListButton = require('./components/OrderedListButton');

var _OrderedListButton2 = _interopRequireDefault(_OrderedListButton);

var _BlockquoteButton = require('./components/BlockquoteButton');

var _BlockquoteButton2 = _interopRequireDefault(_BlockquoteButton);

var _CodeBlockButton = require('./components/CodeBlockButton');

var _CodeBlockButton2 = _interopRequireDefault(_CodeBlockButton);

var _Separator = require('./components/Separator');

var _Separator2 = _interopRequireDefault(_Separator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createInlineToolbarPlugin = function createInlineToolbarPlugin() {
  var config = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  var store = (0, _createStore2.default)({
    isVisisble: false
  });

  var _config$structure = config.structure;
  var structure = _config$structure === undefined ? [_BoldButton2.default, _ItalicButton2.default, _UnderlineButton2.default, _CodeButton2.default] : _config$structure;


  var toolbarProps = {
    store: store,
    structure: structure
  };

  return {
    initialize: function initialize(_ref) {
      var getEditorState = _ref.getEditorState;
      var setEditorState = _ref.setEditorState;

      store.updateItem('getEditorState', getEditorState);
      store.updateItem('setEditorState', setEditorState);
    },
    // Re-Render the text-toolbar on selection change
    onChange: function onChange(editorState) {
      var selection = editorState.getSelection();
      if (selection.getHasFocus() && !selection.isCollapsed()) {
        store.updateItem('isVisible', true);
      } else {
        store.updateItem('isVisible', false);
      }
      return editorState;
    },
    InlineToolbar: (0, _decorateComponentWithProps2.default)(_Toolbar2.default, toolbarProps)
  };
};

exports.default = createInlineToolbarPlugin;
exports.ItalicButton = _ItalicButton2.default;
exports.BoldButton = _BoldButton2.default;
exports.CodeButton = _CodeButton2.default;
exports.UnderlineButton = _UnderlineButton2.default;
exports.HeaderOneButton = _HeaderOneButton2.default;
exports.HeaderTwoButton = _HeaderTwoButton2.default;
exports.HeaderThreeButton = _HeaderThreeButton2.default;
exports.UnorderedListButton = _UnorderedListButton2.default;
exports.OrderedListButton = _OrderedListButton2.default;
exports.BlockquoteButton = _BlockquoteButton2.default;
exports.CodeBlockButton = _CodeBlockButton2.default;
exports.Separator = _Separator2.default;
